PIM-7660: add an extension point in the datagrid (#9212)
akeneo/pim-community-dev: internal_doc/extensibility/product_grid.md

# Product Grid Extensibility 

This documentation explains the different steps to add a new property into the product grid.

> In this example, we will explain here how to add a property to inform when a product has been subscribed to an external system called Suggest Data.

Reviewer: ```suggestion
In this example, we will explain here how to add a property to inform when a product has been subscribed to an external data provider system.
```

Product model rows displayed in the datagrid will not have such date as it is not possible to subscribe to this system (for example purpose).

## Requirements

The property to display in the product datagrid must be indexed into the product and product model Elasticsearch index.

# Configuring the datagrid with the new property
 
In your bundle, add the following content in a new file `Resources/config/product.yml`:

```
datagrid:
    product-grid:
        columns:

>             is_subscribed_to_suggest_data:

Reviewer: ```suggestion
            is_subscribed_to_data_provider:
```

>                 label:         Suggest Data Subscription Date

Reviewer: ```suggestion
                label:         Data Provider Subsription Date
```

>                 data_name:     is_subscribed_to_suggest_data

Reviewer: ```suggestion
                data_name:     is_subscribed_to_data_provider
```

## Add the property into the product rows

Create a class. In this example, all products will have a subscribed date of the current time.

Of course, you will probably execute an SQL query or call an external service to get the fetch the data. 
Please avoid to execute one call per row and try to execute one query to get all data at once, for performance reason.

```
<?php

declare(strict_types=1);

namespace Acme\Bundle\AppBundle\Query;

use Akeneo\Pim\Enrichment\Component\Product\Grid\Query\AddAdditionalProductProperties;
use Akeneo\Pim\Enrichment\Component\Product\Grid\Query\FetchProductAndProductModelRowsParameters;
use Akeneo\Pim\Enrichment\Component\Product\Grid\ReadModel\AdditionalProperty;

class FetchSubscriptionDateForProductRows implements AddAdditionalProductProperties
{
    /**
     * {@inheritdoc}
     */
    public function add(FetchProductAndProductModelRowsParameters $queryParameters, array $rows): array
    {
        $rowsWithAdditionalProperty = [];
        
        $currentDate = new \DateTime();
        $formattedCurrentDate= $now->format('c')
        foreach ($rows as $row) {

>             $property = new AdditionalProperty('is_subscribed_to_suggest_data', $formattedCurrentDate);

Reviewer: ```suggestion
            $property = new AdditionalProperty('is_subscribed_to_data_provider', $formattedCurrentDate);
```

            $rowsWithAdditionalProperty[] = $row->addAdditionalProperty($property);
        }

        return $rowsWithAdditionalProperty;
    }
}
```

Then, you have to declare this service and tag it:
```
    acme.query.fetch_subscription_date_for_product_rows:
        class: 'Acme\Bundle\AppBundle\Query\FetchSubscriptionDateForProductRows'
        tags:
            - { name: akeneo.pim.enrichment.product.grid.add_additional_product_properties }
```

## Add the property in the product model rows

> In this example, a product model does not have any subscription date to the Suggest Data system.

Reviewer: ```suggestion
In this example, a product model does not have any subscription date to the Data Provider system.
```

However, adding the property in the product models is mandatory, because it's declared in the datagrid configuration.

So, the property will be set as `null`.

```
<?php

declare(strict_types=1);

namespace Acme\Bundle\AppBundle\Query;

use Akeneo\Pim\Enrichment\Component\Product\Grid\Query\AddAdditionalProductModelProperties;
use Akeneo\Pim\Enrichment\Component\Product\Grid\Query\FetchProductAndProductModelRowsParameters;
use Akeneo\Pim\Enrichment\Component\Product\Grid\ReadModel\AdditionalProperty;

class FetchSubscriptionDateForProductModelRows implements AddAdditionalProductModelProperties
{
    /**
     * {@inheritdoc}
     */
    public function add(FetchProductAndProductModelRowsParameters $queryParameters, array $rows): array
    {
        $rowsWithAdditionalProperty = [];
        foreach ($rows as $row) {

>             $property = new AdditionalProperty('is_subscribed_to_suggest_data', null);

Reviewer: ```suggestion
            $property = new AdditionalProperty('is_subscribed_to_data_provider', null);
```
